### PR TITLE
@types/react-dom react-dom/server add ServerOptions

### PR DESCRIPTION
**Pull request by 2239559319**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Review by eps1lon — APPROVED** [review]
We're actually missing these on the official docs. Will fix.

**Comment by eps1lon** [comment]
It's not on react.dev. Can you link the PR?

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react-dom ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-dom@latest/))
was missing the following properties:
1. createRoot
1. hydrateRoot
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 156517629b05dea9d762a5d7719da1782e11e391
</p>

File: types/react-dom/server.d.ts [hunk 1]
```diff
@@ -72,22 +76,22 @@ export function renderToString(element: ReactElement): string;
...
-export function renderToStaticMarkup(element: ReactElement): string;
+export function renderToStaticMarkup(element: ReactElement, options?: ServerOptions): string;
 
 /**
  * Similar to `renderToNodeStream`, except this doesn't create extra DOM attributes
  * such as `data-reactid`, that React uses internally. The HTML output by this stream
  * is exactly equal to what `ReactDOMServer.renderToStaticMarkup()` would return.
  */
-export function renderToStaticNodeStream(element: ReactElement): NodeJS.ReadableStream;
+export function renderToStaticNodeStream(element: ReactElement, options?: ServerOptions): NodeJS.ReadableStream;
```
Comment (eps1lon) [inline types/react-dom/server.d.ts:94]:
```suggestion
export function renderToStaticNodeStream(element: ReactElement): NodeJS.ReadableStream;
```
These are not hydrateable

File: types/react-dom/server.d.ts [hunk 2]
```diff
@@ -72,22 +76,22 @@ export function renderToString(element: ReactElement): string;
...
+export function renderToNodeStream(element: ReactElement, options?: ServerOptions): NodeJS.ReadableStream;
 
 /**
  * Similar to `renderToString`, except this doesn't create extra DOM attributes
  * such as `data-reactid`, that React uses internally. This is useful if you want
  * to use React as a simple static page generator, as stripping away the extra
  * attributes can save lots of bytes.
  */
-export function renderToStaticMarkup(element: ReactElement): string;
+export function renderToStaticMarkup(element: ReactElement, options?: ServerOptions): string;
```
Comment (eps1lon) [inline types/react-dom/server.d.ts:87]:
```suggestion
export function renderToStaticMarkup(element: ReactElement): string;
```
not hydrateable

File: types/react-dom/test/react-dom-tests.tsx [hunk 3]
```diff
@@ -85,10 +85,12 @@ describe("ReactDOM", () => {
 describe("ReactDOMServer", () => {
     it("renderToString", () => {
         const content: string = ReactDOMServer.renderToString(React.createElement("div"));
+        ReactDOMServer.renderToString(React.createElement("div"), { identifierPrefix: "react-18-app" });
     });
 
     it("renderToStaticMarkup", () => {
         const content: string = ReactDOMServer.renderToStaticMarkup(React.createElement("div"));
+        ReactDOMServer.renderToStaticMarkup(React.createElement("div"), { identifierPrefix: "react-18-app" });
```
Comment (eps1lon) [inline types/react-dom/test/react-dom-tests.tsx:93]:
```suggestion
        // @ts-expect-error -- not hydrateable
        ReactDOMServer.renderToStaticMarkup(React.createElement("div"), { identifierPrefix: "react-18-app" });
```

**Review by eps1lon — CHANGES_REQUESTED** [review]

**Review by eps1lon — APPROVED** [review]

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react-dom ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-dom@latest/))
was missing the following properties:
1. createRoot
1. hydrateRoot
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 156517629b05dea9d762a5d7719da1782e11e391
</p>

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## react-dom ([<kbd>unpkg</kbd>](https://unpkg.com/browse/react-dom@latest/))
was missing the following properties:
1. createRoot
1. hydrateRoot
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 156517629b05dea9d762a5d7719da1782e11e391
</p>

**Review by rbuckton — APPROVED** [review]

**Comment by eps1lon** [comment]
Ready to merge

**Pull request merged**